src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -8026,22 +8047,25 @@ internal static bool IsReparsePoint(FileSystemInfo fileInfo)
             return fileInfo.Attributes.HasFlag(System.IO.FileAttributes.ReparsePoint);
         }
 
-        internal static bool IsNameSurrogateReparsePoint(string filePath)
+        internal static bool IsReparsePointWithTarget(FileSystemInfo fileInfo)
         {
+            if (!IsReparsePoint(fileInfo))
+            {
+                return false;
+            }
 #if !UNIX
             var data = new WIN32_FIND_DATA();
-            using (SafeFileHandle handle = FindFirstFileEx(filePath, FINDEX_INFO_LEVELS.FindExInfoBasic, ref data, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, 0))
+            using (SafeFileHandle handle = FindFirstFileEx(fileInfo.FullName, FINDEX_INFO_LEVELS.FindExInfoBasic, ref data, FINDEX_SEARCH_OPS.FindExSearchNameMatch, IntPtr.Zero, 0))
             {
-                // Name surrogates are reparse points that point to other named entities local to the filesystem (like symlinks)
+                // Name surrogates (0x20000000) are reparse points that point to other named entities local to the filesystem (like symlinks)
                 // In the case of OneDrive, they are not surrogates and would be safe to recurse into.
                 // This code is equivalent to the IsReparseTagNameSurrogate macro: https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/nf-ntifs-isreparsetagnamesurrogate
-                if (!handle.IsInvalid && (data.dwReserved0 & 0x20000000) == 0)
+                if (!handle.IsInvalid && (data.dwReserved0 & 0x20000000) == 0 && (data.dwReserved0 & IO_REPARSE_TAG_APPEXECLINK) != IO_REPARSE_TAG_APPEXECLINK)
```

Comment:
Why have the two checks of dwReserved0 a different form?

Reply:
The first half is the equivalent of the `IsReparseTagNameSurrogate` macro which explicitly covers the OneDrive scenario.  The addition is specific to this new reparse point type.

Reply:
Sorry, I meant why we do not use
`data.dwReserved0 & IO_REPARSE_TAG_APPEXECLINK) != 0`

Reply:
The IO_REPARSE_TAGs aren't bitflags, so in the case of OneDrive, dwReserved0 has a value of 0x9000601A.  We want this check to succeed so we don't try to access the file and end up downloading it, so the first half passes, the second half resolves to 0x8000001A which also passes (as IO_REPARSE_TAG_APPEXECLINK is 0x8000001B).  If we change it to `& IO_REPARSE_TAG_APPEXECLINK) == 0`, then the second half fails as it's not zero so we end up downloading files from OneDrive.

Reply:
Shouldn't this be just `data.dwReserved0 != IO_REPARSE_TAG_APPEXECLINK` without any bit masking, then?

The current comparison `(data.dwReserved0 & IO_REPARSE_TAG_APPEXECLINK) != IO_REPARSE_TAG_APPEXECLINK` will return false for e.g. `IO_REPARSE_TAG_WCI_TOMBSTONE` = 0xA000001F, or for 0x8000003B if Microsoft later starts using that for something. Because `WinInternalGetLinkType` does not treat those values as AppX reparse tags, I expect `IsReparsePointWithTarget` shouldn't either.

Reply:
I'm certainly no expert on reparse tags, but my concern would be if a reparse tag could have a combination of tags even though they are not unique bit flags.

Reply:
I'm not an expert on reparse tags, either.

Microsoft's sample code compares to `IO_REPARSE_TAG_MOUNT_POINT` = 0xA0000003 without bit masking. <https://github.com/microsoft/Windows-driver-samples/blob/82bd522bb3b893e145d6d73f930edfc04977e4de/filesys/miniFilter/minispy/user/mspyLog.c#L70>

The Reparse Point Tag Request instructions ask the driver developer to specify how the flags in the highest bits of the 32-bit reparse tag should be set. This seems to imply that the developer is not allowed to flip those flags on their own later. <https://docs.microsoft.com/en-us/windows-hardware/drivers/ifs/reparse-point-tag-request>

Some information about the 0x10000000 bit and the `IsReparseTagDirectory` macro, which don't seem to be documented at docs.microsoft.com yet, was posted to the NTFSD list in September 2016. <https://community.osr.com/discussion/comment/279158#Comment_279158>

Reply:
I reviewed all code I found. The tags is absolute values and I think a right way is to do simple comparison.

Reply:
Checking the 0x20000000 flag (IsReparseTagNameSurrogate) can be OK, if you only care about the coarse meaning of the reparse tag and not about the format of the reparse data.

Reply:
@iSazonov do you want to submit a new PR to clean this up?

Reply:
Will do.